### PR TITLE
WIP:  Move purge logic to plugins

### PR DESCRIPTION
**Pull request by amitaibu**
#100

### CODE REVIEW

**Comment by amitaibu** [comment]
Update: Quota purge is now coming from the plugin.

message___site-install

@RoySegall btw, seems that also on 8.x-1.x these settings cannot be saved (due to wrong logic in `\Drupal\message\Form\MessageSettingsForm::submitForm`)

File: src/MessagePurgeBase.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Entity\Query\QueryInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\message\Entity\Message;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Base implementation for OgDeleteOrphans plugins.
```
Comment (RoySegall) [inline src/MessagePurgeBase.php]: `OgDeleteOrphans` :)

File: src/MessagePurgePluginManager.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\Core\Plugin\DefaultPluginManager;
+
+/**
+ * Plugin type manager for MessagePurge plugins.
+ */
+class MessagePurgePluginManager extends DefaultPluginManager {
+
+  /**
+   * Constructs an OgDeleteOrphansPluginManager object.
```
Comment (RoySegall) [inline src/MessagePurgePluginManager.php]: `OgDeleteOrphansPluginManager`

File: src/Form/MessageSettingsForm.php [hunk 3]
```diff
@@ -132,15 +128,15 @@ public function buildForm(array $form, FormStateInterface $form_state) {
    * {@inheritdoc}
    */
   public function submitForm(array &$form, FormStateInterface $form_state) {
-    $config = $this->config('message.settings');
+    parent::submitForm($form, $form_state);
+
+    $config = $this->configFactory->getEditable('message.settings');
```
Comment (amitaibu) [inline src/Form/MessageSettingsForm.php]: @damiankloip got two questions you might know:
1. Why do we need the `getEditable()` here, where a [similar case](https://github.com/amitaibu/og/blob/8.x-1.x/og_ui/src/Form/AdminSettingsForm.php#L152) in OG doesn't?
2. How to deal with plugins that come with their own config? Do we want/ need to allow implementing modules to inject their values in Message's own [purge config](https://github.com/Gizra/message/blob/8.x-1.x/config/install/message.settings.yml#L2)? (which should probably move all under the `purge` key).

Comment (amitaibu) [inline src/Form/MessageSettingsForm.php]: or @jhedstrom if you are around ;)

Comment (amitaibu) [inline src/Form/MessageSettingsForm.php]: just to clarify (2) - I'm looking for an example, to see how/where plugins implemented in different modules hold their config.

Comment (damiankloip) [inline src/Form/MessageSettingsForm.php]: OG extends ConfigFormBase where the coding ethic will actually return you a mutable config object via getEditable ()

Comment (damiankloip) [inline src/Form/MessageSettingsForm.php]: Ha, sorry. CONFIG() METHOD. Not coding ethic :)

File: src/MessagePurgeBase.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      $container->get('config.factory')
+    );
+  }
+
+
+  /**
+   * {@inheritdoc}
+   */
+  public function process(array $ids) {
+    $messages = Message::loadMultiple($ids);
```
Comment (damiankloip) [inline src/MessagePurgeBase.php]: Use the manager to load the storage. Use that here and to delete. Watch out if there are too many though. You can run out of memory pretty easily if you have a lot of process. This is why I wanted an entity iterator in core :)

Comment (amitaibu) [inline src/MessagePurgeBase.php]: yeah, I just throw that one in to get it working. I'll change before setting to needs review :)

File: src/Form/MessageSettingsForm.php (hunk 3 above)
Comment (amitaibu) [inline src/Form/MessageSettingsForm.php]: > OG extends ConfigFormBase

Yeah, this also extends `ConfigFormBase` that's why I'm a bit confused ;)

Comment (amitaibu) [inline src/Form/MessageSettingsForm.php]: btw, never underestimate coding ethics!

File: src/MessagePurgeBase.php (hunk 4 above)
Comment (damiankloip) [inline src/MessagePurgeBase.php]: 👍 Sorry!

File: src/Form/MessageSettingsForm.php (hunk 3 above)
Comment (damiankloip) [inline src/Form/MessageSettingsForm.php]: Too true, my friend. Too true.

Anyway, it is not working the same as the OG form as your config name in `getEditableConfigNames` on this form is not correct, it's `messag.message` needs to be `message.settings` :) With that you will be able to switch back to `config()`.

Comment (amitaibu) [inline src/Form/MessageSettingsForm.php]: @damiankloip you are so knowledgable, I'm starting to have feeling towards you! 💘

File: src/Form/MessageSettingsForm.php [hunk 5]
```diff
@@ -100,21 +100,17 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
 
-    $form['purge']['purge_days'] = [
-      '#type' => 'textfield',
-      '#title' => t('Purge messages older than'),
-      '#description' => t('Maximal message age in days.'),
-      '#default_value' => $config->get('purge_quota'),
-      '#states' => $states,
-    ];
+      // @todo: Set the default value.
+      // $element['#default_value'] = $config->get($plugin_id);
```
Comment (amitaibu) [inline src/Form/MessageSettingsForm.php]: @damiankloip so my plugin's config question is this. Given I have this settings yaml:

``` yaml
# message/config/install/message.settings.yml
purge:
  quota: 0
  days: 0
```

I could set the default value by doing something like `$element['#default_value'] = $config->get('purge.' . $plugin_id);`

However, how would that work if the plugin is defined by another module? -- i.e. it's value isn't in Message's own settings yml file.

Am I taking the right direction here?

Comment (amitaibu) [inline src/Form/MessageSettingsForm.php]: Ok, Tim Plunket gave me the answer - It's very similar to the the image style plugins per plugin config.

**Comment by amitaibu** [comment]
I'm thinking of modeling this after image styles, so each purge plugin will get their own chance to enable/ disable and to edit.

edit_style_large__480x480____site-install-2

**Comment by jhedstrom** [comment]
@RoySegall @amitaibu any updates here? I can potentially pick this up in a week or so if not.

**Comment by RoySegall** [comment]
@jhedstrom No. Don't have time to tackle this one.

**Comment by amitaibu** [comment]
@jhedstrom Would be _great_ if you could pick it up. This PR is just a start, and it's pretty messed up, so you could start a new branch if you prefer.

**Comment by jhedstrom** [comment]
Ok. I should be able to work on this next week.

**Comment by jhedstrom** [comment]
I'm working on this now. I started by rebasing the work done here, and will push to a new PR.

**Pull request closed without merging**